lightning/src/ln/channelmanager.rs

```diff
@@ -1822,6 +1822,48 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 		} else { false }
 	}
 
+	// Fail a list of HTLCs that were just freed from the holding cell. The HTLCs need to be
+	// failed backwards or, if they were one of our outgoing HTLCs, then their failure needs to
+	// be surfaced to the user.
+	fn fail_holding_cell_htlcs(&self, mut htlcs_to_fail: Vec<(HTLCSource, PaymentHash)>, channel_id: [u8; 32]) -> Result<(), MsgHandleErrInternal> {
+		for (htlc_src, payment_hash) in htlcs_to_fail.drain(..) {
+			match htlc_src {
+				HTLCSource::PreviousHopData(HTLCPreviousHopData { .. }) => {
+					let onion_failure_data =
+						match self.channel_state.lock().unwrap().by_id.entry(channel_id) {
+							hash_map::Entry::Occupied(chan_entry) => {
+								if let Ok(upd) = self.get_channel_update(&chan_entry.get()) {
+									let mut res = Vec::with_capacity(8 + 128);
+									res.extend_from_slice(&byte_utils::be16_to_array(upd.contents.flags.into()));
+									res.extend_from_slice(&upd.encode_with_len()[..]);
+									res
+								} else {
+									return Err(MsgHandleErrInternal::send_err_msg_no_close("Failed to find corresponding channel".to_string(), channel_id));
```

Comment:
Hmm, returning early here seems bad - if you have some crazy race where we're handling closing HTLCs for a channel which gets closed before we get here, we may leave HTLCs (on other channels!) un-failed-back.